modules/gdscript/gdscript_compiler.cpp

```diff
@@ -2023,15 +2024,25 @@ Error GDScriptCompiler::_parse_block(CodeGen &codegen, const GDScriptParser::Sui
 
 				gen->write_for_assignment(list);
 
+				if (for_n->second_variable) {
+					second_iterator = codegen.add_local(for_n->second_variable->name, _gdtype_from_datatype(for_n->second_variable->get_datatype(), codegen.script));
+					gen->write_for_dictionary_type_check(second_iterator);
```

Comment:
Type checking at runtime is not needed if the static analyzer has determined that it is always a dictionary.

<details>
<summary>Details</summary>

```gdscript
func test():
    var a: Dictionary
    for k, v in a: # Runtime type check is not needed.
        pass
    var b: Variant # Runtime type check is needed.
    for k, v in b:
        pass
```

```
Disassembling test()
 0: line 2:     var a: Dictionary
 2: construct validated stack(3) = Dictionary()
 7: line 3:     for k, v in a: # Runtime type check is not needed.
 9: assign stack(7) = stack(3)
 12: type test stack(7) is Dictionary  <-- !!!
 15: for-init (typed DICTIONARY) stack(5) in stack(7) counter stack(6) end 36
 20: jump 27
 22: for-loop (typed DICTIONARY) stack(5) in stack(7) counter stack(6) end 36
 27: get keyed validated stack(9) = stack(7)[stack(6)]
 32: line 4:         pass
 34: jump 22
 36: line 5:     var b: Variant # Runtime type check is needed.
 38: line 6:     for k, v in b:
 40: assign stack(7) = stack(4)
 43: type test stack(7) is Dictionary
 46: for-init stack(5) in stack(7) counter stack(6) end 66
 51: jump 58
 53: for-loop stack(7) in stack(7) counter stack(6) end 66
 58: get keyed stack(9) = stack(7)[stack(6)]
 62: line 7:         pass
 64: jump 53
 66: == END ==
```

</details>

Reply:
changed to:
```
if (for_n->list->get_datatype().is_variant()) {
    gen->write_for_dictionary_type_check(second_iterator);
}
```
and it seems correct now:

```
func test():
    var a: Dictionary
    for k, v in a: # Runtime type check is not needed.
        pass
    var b: Variant = [1, 2, 3]
    for u, v in b:  # Runtime type check is needed.
        pass
```

> Disassembling test()
>  0: line 2:     var a: Dictionary
>  2: construct validated stack(3) = Dictionary()
>  7: line 3:     for k, v in a: # Runtime type check is not needed.
>  9: assign stack(7) = stack(3)
>  12: for-init (typed DICTIONARY) stack(5) in stack(7) counter stack(6) end 33
>  17: jump 24
>  19: for-loop (typed DICTIONARY) stack(5) in stack(7) counter stack(6) end 33
>  24: get keyed validated stack(9) = stack(7)[stack(6)]
>  29: line 4:         pass
>  31: jump 19
>  33: line 5:     var b: Variant = [1, 2, 3]
>  35:  make_array stack(10) = [const(1), const(2), const(3)]
>  42: assign stack(4) = stack(10)
>  45: assign stack(10) = null
>  47: line 6:     for u, v in b:  # Runtime type check is needed.
>  49: assign stack(7) = stack(4)
>  52: for_loop_second_var_type_test stack(7) is Dictionary
>  54: for-init stack(5) in stack(7) counter stack(6) end 74
>  59: jump 66
>  61: for-loop stack(7) in stack(7) counter stack(6) end 74
>  66: get keyed stack(9) = stack(7)[stack(6)]
>  70: line 7:         pass
>  72: jump 61
>  74: == END ==